content/courses/advanced-ada/chapters/types.rst

```diff
@@ -1643,19 +1643,913 @@ want to declare objects of this type.
+
+    package Custom_Enumerations is
+
+       type Priority is (Low, Mid, High);
+
+       type Important_Priority is new
+         Priority range Mid .. High;
+
+    end Custom_Enumerations;
+
+    with Ada.Text_IO;         use Ada.Text_IO;
+    with Custom_Enumerations; use Custom_Enumerations;
+
+    procedure Show_Conversion is
+       P  : Priority           := Low;
+       IP : Important_Priority := High;
+    begin
+       P := Priority (IP);
+
+       Put_Line ("P:  "
+                 & P'Image);
+
+       P  := Mid;
+       IP := Important_Priority (P);
+
+       Put_Line ("IP: "
+                 & IP'Image);
+    end Show_Conversion;
+
+In this example, we have the :ada:`Priority` type and the derived type
+:ada:`Important_Priority`. As expected, the conversion works fine when the
+converted value is in the range of the target type. If not, an exception is
+raised:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Enumeration_Type_Conversion
+    :class: ada-run-expect-failure
+
+    with Ada.Text_IO;         use Ada.Text_IO;
+    with Custom_Enumerations; use Custom_Enumerations;
+
+    procedure Show_Conversion is
+       P  : Priority;
+       IP : Important_Priority;
+    begin
+       P  := Low;
+       IP := Important_Priority (P);
+
+       Put_Line ("IP: "
+                 & IP'Image);
+    end Show_Conversion;
+
+In this example, an exception is raised because :ada:`Low` is not in the :ada:`Important_Priority` type's range.
+
+
+Array conversion
+^^^^^^^^^^^^^^^^
+
+Similarly, we can convert between array types. For example, if we have the
+array type :ada:`Integer_Array` and its derived type
+:ada:`Derived_Integer_Array`, we can convert between those array types:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Array_Type_Conversion
+
+    package Custom_Arrays is
+
+       type Integer_Array is array (Positive range <>) of Integer;
+
+       type Derived_Integer_Array is new Integer_Array;
+
+    end Custom_Arrays;
+
+    pragma Ada_2022;
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Custom_Arrays; use Custom_Arrays;
+
+    procedure Show_Conversion is
+       subtype Common_Range is Positive range 1 .. 3;
+
+       AI : Integer_Array (Common_Range);
+       AI_D : Derived_Integer_Array (Common_Range);
+    begin
+       AI_D := [1, 2, 3];
+       AI := Integer_Array (AI_D);
+
+       Put_Line ("AI: "
+                 & AI'Image);
+
+       AI   := [4, 5, 6];
+       AI_D := Derived_Integer_Array (AI);
+
+       Put_Line ("AI_D: "
+                 & AI_D'Image);
+    end Show_Conversion;
+
+Note that both arrays must have the same number of components in order for the
+conversion to be successful. (Sliding is fine, though.) In this example, both
+arrays have the same range: :ada:`Common_Range`.
+
+We can also convert between array types that aren't derived one from the
+other. As long as the components are of the same type, the conversion between
```

Comment:
The index (sub)types are part of the requirements too. Overall:
Component types must be the same type
Index types must be the same or convertible
Dimensionality must be the same
Bounds must be compatible (not necessarily equal)

Reply:
Yes, you're right. (I added these requirements to the paragraph.)